### PR TITLE
Added extra parameters to QueryGranularCertificates

### PR DESCRIPTION
**Pull request by tnickelsen**

### CODE REVIEW

File: src/ProjectOrigin.WalletSystem.Server/Database/Scripts/202306151348-create-attributes-table.sql
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Database/Scripts/202306151348-create-attributes-table.sql:1]: I think there should be a UNIQUE constrait on CertificateId RegistryId and KeyAtr, so no Key can be specificed twice for a certificate.

File: src/ProjectOrigin.WalletSystem.Server/Models/Certificate.cs [hunk 1]
```diff
@@ -1,6 +1,35 @@
 using System;
+using System.Collections.Generic;
 
 namespace ProjectOrigin.WalletSystem.Server.Models;
 
-public record Certificate(Guid Id,
-                          Guid RegistryId);
+public enum GranularCertificateType
+{
+    Invalid = 0,
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Models/Certificate.cs:8]: I do not think we need this type written here, even if it is exists on the types in protobuf, simply let c# throw an invalid cast exception.

File: src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs [hunk 2]
```diff
@@ -17,10 +29,35 @@ public GranularCertificate ToProto()
...
+    private V1.GranularCertificateType ToDto(GranularCertificateType type)
+    {
+        if (type == GranularCertificateType.Production)
+            return V1.GranularCertificateType.Production;
+
+        if (type == GranularCertificateType.Consumption)
+            return V1.GranularCertificateType.Consumption;
+
+        if (type == GranularCertificateType.Invalid)
+            return V1.GranularCertificateType.Invalid;
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs:59]:
```suggestion
```

Remove and let the exception be thrown since it is invalid.

File: src/ProjectOrigin.WalletSystem.IntegrationTests/Repositories/AbstractRepositoryTests.cs [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 using ProjectOrigin.WalletSystem.Server.HDWallet;
 using ProjectOrigin.WalletSystem.Server.Models;
 using ProjectOrigin.WalletSystem.Server.Repositories;
 using System;
+using System.Collections.Generic;
 using System.Data;
 using System.Threading.Tasks;
+using ProjectOrigin.WalletSystem.Server.Extensions;
 using Xunit;
+using ProjectOrigin.WalletSystem.Server.Models;
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.IntegrationTests/Repositories/AbstractRepositoryTests.cs:14]:
```suggestion
```

dupplicate

File: src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs [hunk 4]
```diff
@@ -1,11 +1,23 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Google.Protobuf.WellKnownTypes;
 using ProjectOrigin.Register.V1;
 using ProjectOrigin.WalletSystem.V1;
 
 namespace ProjectOrigin.WalletSystem.Server.Models;
 
-public record CertificateViewModel(Guid Id, string Registry, long Quantity)
+public class CertificateViewModel
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs:10]: How come you have changed it to a class, you already instantiate the attribute and slices with empty lists, so these could easily be filled without the set?

File: src/ProjectOrigin.WalletSystem.Server/ProjectOrigin.WalletSystem.Server.csproj [hunk 5]
```diff
@@ -6,8 +6,13 @@
     <ImplicitUsings>disable</ImplicitUsings>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="Database\Scripts\202306151348-create-attributes-table.sql" />
+  </ItemGroup>
+
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/ProjectOrigin.WalletSystem.Server.csproj:12]:
```suggestion
```

not needed?

File: src/ProjectOrigin.WalletSystem.Server/Models/SliceViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+            var b = obj as SliceViewModel;
+            return SliceId.Equals(b!.SliceId);
+        }
+
+        public override int GetHashCode()
+        {
+            return SliceId.GetHashCode() * Quantity.GetHashCode();
+        }
+    }
+}
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Models/SliceViewModel.cs:22]: Is mutability of properties required? 
Otherwise record could remove the need for the equability code and simplify the model greatly.

File: src/ProjectOrigin.WalletSystem.Server/Repositories/CertificateRepository.cs
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Repositories/CertificateRepository.cs:1]: Looking at this highlights we probably need to look at filter keys in the near future, this basically joins everything in the database to be able to answer.
This means that all the rows are returned for each different attribute, maybe we should have a look at this.

@duizer do you know of other ways to do this in dapper?

File: src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using Google.Protobuf.WellKnownTypes;
+using NBitcoin.RPC;
```
Comment (MartinSchmidt) [inline src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs]: why is this included now?

**Review by MartinSchmidt — CHANGES_REQUESTED** [review]

File: src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs (hunk 7 above)
Comment (tnickelsen) [reply src/ProjectOrigin.WalletSystem.Server/Models/CertificateViewModel.cs]: Think you are looking at an outdated version

**Review by MartinSchmidt — APPROVED** [review]
